### DISCUSSION TITLE
How to parse MD tags in CRAM files

### DISCUSSION THREAD

**Discussion by Lilas-w** [GMOD/jbrowse-components#3327] [Q&A] [answered]
Hello, I have BAM files and CRAM files with the same content. For BAM files, JBrowse code can parse MD tags in `feature.record.data` like this:

![bam](https://user-images.githubusercontent.com/81499006/201566333-c0d3fc2e-0ea0-4129-9f24-3db7a7cac0ab.png)

But for CRAM files, there are no MD tags in `feature.record`:

![cram](https://user-images.githubusercontent.com/81499006/201566388-93200a6d-0592-4e9a-bbeb-dc97857303b6.png)

How could I get MD tags in CRAM files just like in BAM files by writing some code? 

I need to filter specific mismatch bases and match bases using MD strings.

Thanks!

**Comment by cmdcolin** [comment, marked as answer]
I'd try using 

`feature.get('mismatches')`

the mismatches is an array of objects that jbrowse computes that tells you where the mismatches in the reads are. this avoids having to fiddle with MD tag, which is quite finicky anyways

feature.get('mismatches') will return an object that is the same for BAM and CRAM data

using the "feature.get" function is also generally recommended over poking into the objects like feature.record directly, as the internals of the read objects could change (unlikely, but it's just sort of a "private api" you could say)

**Reply by Lilas-w** [reply]
Maybe our definition of mismatches is a little different. Let me describe the question more clear. When using `feature.get('mismatches')`, I get this in BAM files:

![bam-mismatches](https://user-images.githubusercontent.com/81499006/201588990-5cbffd8c-e4d5-4186-8cd6-3398da7ee032.png)

And get this in CRAM files:

![cram-mismatches](https://user-images.githubusercontent.com/81499006/201589855-d339c3ad-0ad4-447b-94eb-82ddf59a3566.png)

However, the mismatches array with eight items isn't what I need. Take the md tag('20C1C6C3C3C44A20T1T45') in the first screenshot as an example, I need to iterate the digital part in the md tag and set some if conditions to filter specific match bases in the sequence, not only the eight letters(mismatches). Actually I need "matches". In this situation, using MD tags is very convenient to me. My code run successfully using BAM files, so I really want to know how to parse MD tags in CRAM files.

Thanks!

**Reply by Lilas-w** [reply]
Or could you tell me where is the source code for parsing MD tags in BAM files? Thanks!

**Reply by cmdcolin** [reply]
can you describe exactly what you are trying to achieve?

I will just add a technical note to explain my confusion:

it is an sort of difficult to explain, but as far as I know MD is irrelevant to CRAM. my reasoning is that  MD tag is really only useful to tell you where the mismatches are in the absence of a corresponding FASTA with a BAM file. You cannot tell where SNPs are in a BAM file that lack MD without comparing all the CIGAR match sequences with the corresponding locations in a FASTA file. But, once you have added MD tag to that BAM file, you can then infer SNP positions without looking at a FASTA file.

CRAM does not have this same need, and so CRAM files generally lacks MD. I think sometimes it can have it, but it would be wasted space in the file as far as I know.

**Reply by Lilas-w** [reply]
We would like to coloring bases on reads from methylation-like sequencing data: coloring converted C (C -> T) as red, and un-converted C as blue, and all others as grey. To get un-converted C, I use MD tags.

Or another question, which is also a problem I have to solve, how could I parse any tags in BAM and CRAM files (including newly added custom tags) by writing some code? Do I need to change the code in the bam and cram packages imported by JBrowse?

Thanks!

**Reply by cmdcolin** [reply]
I don't think MD tag will tell you where "unconverted" (non-SNP) positions would be. The MD tag tells you what the reference base was at the position where there was a SNP. 

Just to clarify, you are trying to plot WGBS data? That would be a good feature for jbrowse 2 to have.  You can also look at our code for plotting methylation using the MM tag for long reads, but it might be more complicated than is needed for your purposes. It explicitly fetches the reference sequence separately, you can see in PileupRenderer it looks at `regionSequence`.


One alternative strategy to MD and that does not use regionSequence: you can get the read sequence using feature.get('seq'), and any position of the read sequence that is not a mismatch, but is a C, **AND** is not inside an insert or clipping (have to inspect the CIGAR for this, with feature.get('CIGAR')), would be an "unconverted" C. I use code like this to parse CIGAR https://cmdcolin.github.io/posts/2022-02-06-sv-sam#appendix-b---cigar-parsing

**Reply by Lilas-w** [reply]
Thanks for your reply again. Our data isn't WGBS data. I actually used MD tags with feature.get('seq') to filter out mismatches, then got C bases from digital parts. I can rewrite it using other way, like you mentioned. Thank you for the reminder to use CIGAR, although our data doesn't have many insertions or clips now, taking it into account will make our code more applicable.

But we still have a question that, in general, if we add some custom tags with custom names to BAM and CRAM files in the future, how could I parse those new tags to items of `tags` or `feature.record` array and get them? I guess `getTag` function may help, but it doesn't work for me on NM and MD tag in CRAM files. Maybe these two tags are different from other custom tags.

I found a demo https://jbrowse.org/code/jb2/v2.1.2/?config=test_data%2Fconfig_demo.json&session=local-2CNgLeo5- with tags like this:

![custom tags](https://user-images.githubusercontent.com/81499006/201857640-de134f70-c07a-4f2d-a496-fe9661ecbacc.png)

I have already read `parseCigar`, `cigarToMismatches`, `mdToMismatches` function in alignments\src\BamAdapter\MismatchParser.ts, and `_get_CIGAR` function in alignments\src\CramAdapter\CramSlightlyLazyFeature.ts, but still feel confused.

Thanks!

**Reply by cmdcolin** [reply]
Hi sorry for the delay @Lilas-w 

One thing is sort of related to this is mentioned in the CRAM spec https://samtools.github.io/hts-specs/CRAMv3.pdf

```
The MD and NM auxiliary tags store the differences (an edit string) between the sequence and the reference
along with the number of mismatches. These may optionally be created on-the-fly during reference-based
sequence reconstruction and should match the description provided in the SAMtags document. An encoder
may decide to store these verbatim when no reference is used or where the automatically constructed values
differ to the input data.
Note there is no mechanism to describe which records have MD/NM present and which do not. If this is deemed
important, the only recourse is to store all MD and NM verbatim and to request that the decoding software
does not automatically generate its own for records that have no stored MD and NM tags.
```

This means that often times MD and NM are not explicitly stored in the CRAM file. some decoders, e.g. samtools, might recreate this on the fly while serializing to SAM format maybe. the same thing happens with CIGAR. the CIGAR string is not stored in CRAM, it uses a different representation. JBrowse has tools to reconstruct the CIGAR string from CRAM (this is the _get_CIGAR function in CramSlightlyLazyFeature). then the other functions like the cigarToMismatches converts to the mismatches data structure.  JBrowse could potentially reconstruct the NM tag for CRAM files when one does not exist like the spec says potentially, but you can also manually compute it e.g. by looking at the mismatches in the feature.get('mismatches') array.

If you have any more details on what you want to do I might be able to aid more :) it sounds like a great project

**Reply by Lilas-w** [reply]
Thanks for your detailed reply. 

Do you mean that all tags that a general CRAM file doesn't stored can't be parsed (such as MD, NM, and especially **custom tags** that other CRAM files in the world don't have) ? But in this demo, many **custom tags** in CRAM file could be parsed. How this demo did this?
https://jbrowse.org/code/jb2/v2.1.2/?config=test_data%2Fconfig_demo.json&session=local-2CNgLeo5- 

![cram](https://user-images.githubusercontent.com/81499006/202629212-c15c36d5-3ad8-4871-b1af-01d56039c412.png)

 What I'm trying to achieve is adding new **custom tags** to CRAM/BAM file (in our situation they have the same content) and getting its value by JBrowse code.

Thanks!

**Reply by cmdcolin** [reply]
Custom tags will work, you can add these to your file and jbrowse will read the tag data. The MD and NM tags are a special case where they are not stored in CRAM.

To read the tag, just note that the feature objects for BAM reads and CRAM are slightly different. To smooth over the difference between them we use the getTag function. You can look at the code for the getTag function to see (basically the difference is on bam reads it's just feature.get('tagname') and on cram it's feature.get('tags').tagname but calling getTag(feature, 'tagname') works on either bam or cram reads). 

For more technical detail, the reason why the BAM and CRAM features are different like this is that the BAM reads are more lazily decoded than CRAM reads, which is a speed optimization which means the BAM code doesn't have to convert the binary tag data into JS objects until actually requested which helps BAM speed.
